File: Algorithms/C++/Linked_List/143_Reorder_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        if (head == NULL) {
            return;
        }

        ListNode* mid = get_middle(head);
        ListNode* right_reversed = reverse(mid->next);
        mid->next = NULL;

        merge(head, right_reversed);
    }
private:
    ListNode* get_middle(ListNode* head) {
        if (head == NULL) {
            return NULL;
        }

        ListNode* fast = head->next;
        ListNode* slow = head;

        while (fast != NULL && fast->next != NULL) {
            fast = fast->next->next;
            slow = slow->next;
        }

        return slow;
    }

    ListNode* reverse(ListNode* head) {
        ListNode* prev = NULL;

        while (head != NULL) {
            ListNode* tmp = head->next;
            head->next = prev;
            prev = head;
            head = tmp;
        }

        return prev;
    }

    void merge(ListNode* head1, ListNode* head2) {
        ListNode dummy(0);
        ListNode* node = &dummy;

        for (int i = 1; head1 != NULL && head2 != NULL; i++) {
            if (i % 2 != 0) {
                node->next = head1;
                head1 = head1->next;
            } else {
                node->next = head2;
                head2 = head2->next;
            }
            node = node->next;
        }

        if (head1 != NULL) {
            node->next = head1;
        }
        if (head2 != NULL) {
            node->next = head2;
        }
    }
};

File: Algorithms/JavaScript/Medium/143_Reorder_List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
    if (head === null) {
        return
    }
    
    var middle = getMiddle(head)
    var list1 = head
    var list2 = middle.next
    middle.next = null
    
    list2 = reverseList(list2)
    zipList(list1, list2)
};

var getMiddle = function(head) {
    if (head === null) {
        return null
    }
    
    var slow = head, fast = head.next
    while (fast !== null && fast.next !== null) {
        slow = slow.next
        fast = fast.next.next
    }
    
    return slow
}

var reverseList = function(head) {
    var previous = null
    
    while (head !== null) {
        var nextNode = head.next
        head.next = previous
        previous = head
        head = nextNode
    }
    
    return previous
}

var zipList = function(head1, head2) {
    var dummy = new ListNode(0)
    var node = dummy, i = 0
    while (head1 !== null && head2 !== null) {
        if (i % 2 === 0) {
            node.next = head1
            head1 = head1.next
        } else {
            node.next = head2
            head2 = head2.next
        }
        node = node.next
        i++
    }
    
    if (head1 !== null) {
        node.next = head1
    }
    if (head2 !== null) {
        node.next = head2
    }
    
    return dummy.next
}
